File: iconex/src/main/java/foundation/icon/iconex/view/ui/create/CreateWalletStep2Fragment.java
package foundation.icon.iconex.view.ui.create;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.ScrollView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import foundation.icon.ICONexApp;
import foundation.icon.MyConstants;
import foundation.icon.iconex.R;
import foundation.icon.iconex.util.PasswordValidator;
import foundation.icon.iconex.util.Utils;
import foundation.icon.iconex.wallet.Wallet;
import foundation.icon.iconex.wallet.WalletEntry;
import foundation.icon.iconex.widgets.TTextInputLayout;
import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;
import loopchain.icon.wallet.core.Constants;
import loopchain.icon.wallet.service.crypto.KeyStoreUtils;

import static foundation.icon.iconex.util.PasswordValidator.checkPasswordMatch;

public class CreateWalletStep2Fragment extends Fragment implements View.OnClickListener {
    private static final String TAG = CreateWalletStep2Fragment.class.getSimpleName();

    private OnStep2Listener mListener;
    private CreateWalletViewModel vm;
    private Disposable disposable;

    public static CreateWalletStep2Fragment newInstance() {
        return new CreateWalletStep2Fragment();
    }

    private ScrollView scroll;
    private TTextInputLayout inputAlias, inputPwd, inputCheck;
    private ViewGroup layoutButtons;
    private Button btnBack, btnNext;
    private ProgressBar progress;

    private String beforeStr, beforePwd;

    private final int OK = 0;
    private final int ALIAS_DUP = 1;
    private final int ALIAS_EMPTY = 2;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        vm = ViewModelProviders.of(Objects.requireNonNull(getActivity())).get(CreateWalletViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.layout_create_wallet_step2, container, false);
        initView(v);

        return v;
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);

        if (context instanceof OnStep2Listener) {
            mListener = (OnStep2Listener) context;
        } else {
            throw new RuntimeException(context + " must implement OnStep2Listener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        if (disposable != null
                && !disposable.isDisposed())
            disposable.dispose();

        mListener = null;
    }

    private void initView(View v) {
        scroll = v.findViewById(R.id.scroll);
        layoutButtons = v.findViewById(R.id.layout_buttons);
        inputAlias = v.findViewById(R.id.input_alias);
        inputAlias.disableCopyPaste();
        inputAlias.setOnFocusChangedListener(new TTextInputLayout.OnMyFocusChangedListener() {
            @Override
            public void onFocused() {
            }

            @Override
            public void onReleased() {
                if (checkAlias(inputAlias.getText()) == ALIAS_DUP) {
                    inputAlias.setError(true, getString(R.string.duplicateWalletAlias));
                } else {
                    inputAlias.setError(false, null);
                }
            }
        });
        inputAlias.setOnTextChangedListener(new TTextInputLayout.OnTextChanged() {
            @Override
            public void onChanged(@NotNull CharSequence s) {
                if (s.length() > 0) {
                    if (s.toString().trim().isEmpty()) {
                        inputAlias.setText("");
                    } else if (s.charAt(0) == ' ') {
                        inputAlias.setText(beforeStr);
                    } else {
                        if (Utils.checkByteLength(s.toString()) > 16) {
                            inputAlias.setText(beforeStr);
                        } else {
                            beforeStr = s.toString();
                        }
                    }
                } else {
                    btnNext.setEnabled(false);
                }
            }
        });
        inputAlias.setOnKeyPreImeListener(onKeyPreIme);

        inputPwd = v.findViewById(R.id.input_pwd);
        inputPwd.disableCopyPaste();
        inputPwd.setOnFocusChangedListener(new TTextInputLayout.OnMyFocusChangedListener() {
            @Override
            public void onFocused() {
            }

            @Override
            public void onReleased() {
                int result = PasswordValidator.validatePassword(inputPwd.getText());
                switch (result) {
                    case PasswordValidator.LEAST_8:
                        inputPwd.setError(true, getString(R.string.errAtLeast));
                        break;
                    case PasswordValidator.NOT_MATCH_PATTERN:
                        inputPwd.setError(true, getString(R.string.errPasswordPatternMatch));
                        break;

                    case PasswordValidator.HAS_WHITE_SPACE:
                        inputPwd.setError(true, getString(R.string.errWhiteSpace));
                        break;

                    case PasswordValidator.SERIAL_CHAR:
                        inputPwd.setError(true, getString(R.string.errSerialChar));
                        break;
                    default:
                        inputPwd.setError(false, null);
                }
            }
        });
        inputPwd.setOnTextChangedListener(new TTextInputLayout.OnTextChanged() {
            @Override
            public void onChanged(@NotNull CharSequence s) {
                if (s.length() > 0) {
                    if (s.charAt(s.length() - 1) == ' ') {
                        inputPwd.setText(s.subSequence(0, s.length() - 1).toString());
                    } else if (s.toString().contains(" ")) {
                        inputPwd.setText(beforePwd);
                    } else {
                        beforePwd = s.toString();
                    }
                } else {
                    btnNext.setEnabled(false);
                }
            }
        });
        inputPwd.setOnKeyPreImeListener(onKeyPreIme);

        inputCheck = v.findViewById(R.id.input_pwd_check);
        inputCheck.disableCopyPaste();
        inputCheck.setOnFocusChangedListener(new TTextInputLayout.OnMyFocusChangedListener() {
            @Override
            public void onFocused() {
            }

            @Override
            public void onReleased() {
                if (inputCheck.getText().isEmpty()) {
                    btnNext.setEnabled(false);
                } else {
                    if (inputCheck.getText().isEmpty()) {
                        inputCheck.setError(true, getString(R.string.errPasswordNotMatched));
                    } else {
                        boolean result = PasswordValidator.checkPasswordMatch(inputCheck.getText(),
                                inputCheck.getText());
                        if (!result) {
                            inputCheck.setError(true, getString(R.string.errPasswordNotMatched));
                        } else {
                            inputCheck.setError(false, null);
                        }
                    }
                }
            }
        });
        inputCheck.setOnTextChangedListener(new TTextInputLayout.OnTextChanged() {
            @Override
            public void onChanged(@NotNull CharSequence s) {
                if (s.length() == 0) btnNext.setEnabled(false);
            }
        });
        inputCheck.setOnEditorActionListener(new TTextInputLayout.OnEditorAction() {
            @Override
            public void onDone() {
                setNextEnable(inputAlias.getText(), inputPwd.getText(), inputCheck.getText());
            }
        });
        inputCheck.setOnKeyPreImeListener(onKeyPreIme);

        btnNext = v.findViewById(R.id.btn_next);
        btnNext.setOnClickListener(this);
        btnBack = v.findViewById(R.id.btn_back);
        btnBack.setOnClickListener(this);
        progress = v.findViewById(R.id.progress);
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btn_next:
                btnNext.setEnabled(false);
                progress.setVisibility(View.VISIBLE);

                disposable = Observable.create(new ObservableOnSubscribe<String[]>() {
                    @Override
                    public void subscribe(ObservableEmitter<String[]> emitter) throws Exception {
                        emitter.onNext(createWallet());
                        emitter.onComplete();
                    }
                }).subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableObserver<String[]>() {
                            @Override
                            public void onNext(String[] result) {
                                Wallet wallet = new Wallet();
                                wallet.setAlias(Utils.strip(inputAlias.getText()));
                                wallet.setAddress(result[0]);
                                vm.setPrivateKey(result[1]);
                                wallet.setKeyStore(result[2]);
                                wallet.setCoinType(vm.getCoin().getValue().getSymbol().toUpperCase());

                                List<WalletEntry> entries = new ArrayList<>();
                                WalletEntry coin = new WalletEntry();
                                coin.setType(MyConstants.TYPE_COIN);
                                coin.setAddress(result[0]);

                                if (wallet.getCoinType().equals(Constants.KS_COINTYPE_ICX)) {
                                    coin.setName(MyConstants.NAME_ICX);
                                    coin.setSymbol(Constants.KS_COINTYPE_ICX);
                                } else {
                                    coin.setName(MyConstants.NAME_ETH);
                                    coin.setSymbol(Constants.KS_COINTYPE_ETH);
                                }

                                entries.add(coin);
                                wallet.setWalletEntries(entries);

                                vm.setWallet(wallet);

                                mListener.onStep2Done();
                            }

                            @Override
                            public void onError(Throwable e) {

                            }

                            @Override
                            public void onComplete() {
                                btnNext.setEnabled(true);
                                progress.setVisibility(View.GONE);
                            }
                        });
                break;

            case R.id.btn_back:
                mListener.onStep2Back();
                break;
        }
    }

    private int checkAlias(String target) {
        String alias = Utils.strip(target);
        if (alias.isEmpty())
            return ALIAS_EMPTY;

        for (Wallet wallet : ICONexApp.wallets) {
            if (wallet.getAlias().equals(alias))
                return ALIAS_DUP;
        }

        return OK;
    }

    private void setNextEnable(String alias, String pwd, String checkPwd) {
        int aliasValidate = 0;
        int pwdValidate = 0;
        boolean matched = true;

        if (!inputAlias.toString().isEmpty()) {
            aliasValidate = checkAlias(alias);
            switch (aliasValidate) {
                case ALIAS_DUP:
                    inputAlias.setError(true, getString(R.string.duplicateWalletAlias));
                    break;
                default:
                    inputAlias.setError(false, null);
            }
        } else {
            aliasValidate = ALIAS_EMPTY;
            btnNext.setEnabled(false);
            return;
        }

        if (!inputPwd.getText().isEmpty()) {
            pwdValidate = PasswordValidator.validatePassword(pwd);
            if (pwdValidate != 0) {
                switch (pwdValidate) {
                    case PasswordValidator.LEAST_8:
                        inputPwd.setError(true, getString(R.string.errAtLeast));
                        break;
                    case PasswordValidator.NOT_MATCH_PATTERN:
                        inputPwd.setError(true, getString(R.string.errPasswordPatternMatch));
                        break;

                    case PasswordValidator.HAS_WHITE_SPACE:
                        inputPwd.setError(true, getString(R.string.errWhiteSpace));
                        break;

                    case PasswordValidator.SERIAL_CHAR:
                        inputPwd.setError(true, getString(R.string.errSerialChar));
                        break;
                    default:
                        inputPwd.setError(false, null);
                }
            }
        } else {
            pwdValidate = PasswordValidator.EMPTY;
            btnNext.setEnabled(false);
            return;
        }

        if (!inputCheck.getText().isEmpty()) {
            matched = checkPasswordMatch(pwd, checkPwd);
            if (!matched) {
                inputCheck.setError(true, getString(R.string.errPasswordNotMatched));
            } else {
                inputCheck.setError(false, null);
            }
        } else
            matched = false;


        if (aliasValidate == OK && matched && (pwdValidate == PasswordValidator.OK))
            btnNext.setEnabled(true);
        else
            btnNext.setEnabled(false);
    }

    private String[] createWallet() {
        if (vm.getCoin().getValue() == MyConstants.Coin.ICX) {
            return KeyStoreUtils.generateICXKeystore(inputPwd.getText());
        } else {
            return KeyStoreUtils.generateEtherKeystore(inputPwd.getText());
        }
    }

    private TTextInputLayout.OnKeyPreIme onKeyPreIme = new TTextInputLayout.OnKeyPreIme() {
        @Override
        public void onDone() {
            setNextEnable(inputAlias.getText(), inputPwd.getText(), inputCheck.getText());
        }
    };

    public interface OnStep2Listener {
        void onStep2Done();

        void onStep2Back();
    }
}

File: iconex/src/main/java/foundation/icon/iconex/view/ui/create/CreateWalletStep2Fragment.kt
package foundation.icon.iconex.view.ui.create

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ProgressBar
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import foundation.icon.ICONexApp
import foundation.icon.MyConstants
import foundation.icon.iconex.R
import foundation.icon.iconex.util.PasswordValidator
import foundation.icon.iconex.util.PasswordValidator.checkPasswordMatch
import foundation.icon.iconex.util.Utils
import foundation.icon.iconex.wallet.Wallet
import foundation.icon.iconex.wallet.WalletEntry
import foundation.icon.iconex.widgets.TTextInputLayout
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import loopchain.icon.wallet.core.Constants
import loopchain.icon.wallet.service.crypto.KeyStoreUtils
import java.util.*

class CreateWalletStep2Fragment : Fragment(), View.OnClickListener {

    private var mListener: OnStep2Listener? = null

    private lateinit var inputAlias: TTextInputLayout
    private lateinit var inputPwd: TTextInputLayout
    private lateinit var inputCheck: TTextInputLayout

    private lateinit var btnBack: Button
    private lateinit var btnNext: Button
    private lateinit var progress: ProgressBar

    private var beforeStr: String = ""
    private var beforePwd: String = ""
    private var beforeCheck: String = ""

    private val OK = 0
    private val ALIAS_DUP = 1
    private val ALIAS_EMPTY = 2

    private val vm: CreateWalletViewModel by lazy {
        ViewModelProviders.of(activity!!).get(CreateWalletViewModel::class.java)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.layout_create_wallet_step2, container, false)
        initView(view)

        return view
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnStep2Listener) {
            mListener = context
        } else {
            throw RuntimeException("$context must implement OnStep2Listener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    private fun initView(v: View) {
        inputAlias = v.findViewById(R.id.input_alias)
        inputAlias.setOnFocusChangedListener(object : TTextInputLayout.OnFocusReleased {
            override fun onReleased() {
                when (checkAlias(inputAlias.getText())) {
                    ALIAS_DUP -> inputAlias.setError(true, getString(R.string.duplicateWalletAlias))
                    else -> inputAlias.setError(false, null)
                }
            }
        })
        inputAlias.setOnTextChangedListener(object : TTextInputLayout.OnTextChanged {
            override fun onChanged(s: CharSequence) {
                if (s.isNotEmpty()) {
                    if (s.toString().trim { it <= ' ' }.isEmpty()) {
                        inputAlias.setText("")
                    } else if (s[0] == ' ') {
                        inputAlias.setText(beforeStr)
                    } else {
                        if (Utils.checkByteLength(s.toString()) > 16) {
                            inputAlias.setText(beforeStr)
                        } else {
                            beforeStr = s.toString()
                        }
                    }
                } else {
                    btnNext.isEnabled = false
                }
            }
        })
        inputAlias.setOnKeyPreImeListener(onKeyPreIme)

        inputPwd = v.findViewById(R.id.input_pwd)
        inputPwd.setOnFocusChangedListener(object : TTextInputLayout.OnFocusReleased {
            override fun onReleased() {
                val result = PasswordValidator.validatePassword(inputPwd.getText())
                when (result) {

                    PasswordValidator.LEAST_8 -> inputPwd.setError(true, getString(R.string.errAtLeast))

                    PasswordValidator.NOT_MATCH_PATTERN -> inputPwd.setError(true, getString(R.string.errPasswordPatternMatch))

                    PasswordValidator.HAS_WHITE_SPACE -> inputPwd.setError(true, getString(R.string.errWhiteSpace))

                    PasswordValidator.SERIAL_CHAR -> inputPwd.setError(true, getString(R.string.errSerialChar))

                    else -> inputPwd.setError(false, null)
                }
            }
        })
        inputPwd.setOnTextChangedListener(object : TTextInputLayout.OnTextChanged {
            override fun onChanged(s: CharSequence) {
                if (s.isNotEmpty()) {
                    if (s[s.length - 1] == ' ') {
                        inputPwd.setText(s.subSequence(0, s.length - 1).toString())

                    } else if (s.toString().contains(" ")) {
                        inputPwd.setText(beforePwd)
                    } else {
                        beforePwd = s.toString()
                    }
                } else {
                    btnNext.isEnabled = false
                }
            }
        })
        inputPwd.setOnKeyPreImeListener(onKeyPreIme)

        inputCheck = v.findViewById(R.id.input_pwd_check)
        inputCheck.setOnTextChangedListener(object : TTextInputLayout.OnTextChanged {
            override fun onChanged(s: CharSequence) {
                if (s.isNotEmpty()) {
                    if (s[s.length - 1] == ' ') {
                        inputCheck.setText(s.subSequence(0, s.length - 1).toString())
                    } else if (s.toString().contains(" ")) {
                        inputCheck.setText(beforeCheck)
                    } else {
                        beforeCheck = s.toString()
                    }
                } else {
                    btnNext.isEnabled = false
                }
            }
        })
        inputCheck.setOnKeyPreImeListener(onKeyPreIme)
        inputCheck.setOnEditorActionListener(object : TTextInputLayout.OnEditorAction {
            override fun onDone() {
                setNextEnable(inputAlias.getText(), inputPwd.getText(), inputCheck.getText())
            }
        })

        btnBack = v.findViewById(R.id.btn_back)
        btnBack.setOnClickListener(this)
        btnNext = v.findViewById(R.id.btn_next)
        btnNext.setOnClickListener(this)

        progress = v.findViewById(R.id.progress)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_next -> {
                btnNext.isEnabled = false
                progress.visibility = View.VISIBLE

                Observable.just(0)
                        .map { createWallet() }.subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(object : Observer<Array<String>> {
                            override fun onComplete() {
                                progress.visibility = View.GONE
                            }

                            override fun onSubscribe(d: Disposable) {
                            }

                            override fun onNext(t: Array<String>) {
                                val wallet = Wallet()
                                wallet.alias = inputAlias.getText()
                                wallet.address = t[0]
                                vm.setPrivateKey(t[1])
                                wallet.keyStore = t[2]
                                wallet.coinType = vm.getCoinType().value!!.type

                                val entries = ArrayList<WalletEntry>()
                                val coin = WalletEntry()
                                coin.type = MyConstants.TYPE_COIN
                                coin.address = t[0]

                                if (wallet.coinType == Constants.KS_COINTYPE_ICX) {
                                    coin.name = MyConstants.NAME_ICX
                                    coin.symbol = Constants.KS_COINTYPE_ICX
                                } else {
                                    coin.name = MyConstants.NAME_ETH
                                    coin.symbol = Constants.KS_COINTYPE_ETH
                                }

                                entries.add(coin)
                                wallet.walletEntries = entries

                                vm.setWallet(wallet)

                                mListener!!.onStep2Done()
                            }

                            override fun onError(e: Throwable) {
                                e.printStackTrace()
                            }
                        })
            }

            R.id.btn_back -> mListener!!.onStep2Back()
        }
    }

    private var onKeyPreIme = object : TTextInputLayout.OnKeyPreIme {
        override fun onDone() {
            setNextEnable(inputAlias.getText(),
                    inputPwd.getText(),
                    inputCheck.getText())
        }
    }

    private fun checkAlias(target: String): Int {
        val alias = Utils.strip(target)
        if (alias.isEmpty())
            return ALIAS_EMPTY

        for (info in ICONexApp.mWallets) {
            if (info.alias == alias) {
                return ALIAS_DUP
            }
        }

        return OK
    }

    private fun setNextEnable(alias: String, pwd: String, checkPwd: String) {
        val aliasValidate: Int
        val pwdValidate: Int
        val matched: Boolean

        if (inputAlias.getText().isNotEmpty()) {
            aliasValidate = checkAlias(alias)
            when (aliasValidate) {
                ALIAS_EMPTY -> inputAlias.setError(true, getString(R.string.errAliasEmpty))

                ALIAS_DUP -> inputAlias.setError(true, getString(R.string.duplicateWalletAlias))
                else -> inputAlias.setError(false, null)
            }
        } else {
            inputAlias.setError(true, getString(R.string.errAliasEmpty))
            btnNext.isEnabled = false
            return
        }

        if (inputPwd.getText().isNotEmpty()) {
            pwdValidate = PasswordValidator.validatePassword(pwd)
            if (pwdValidate != 0) {
                when (pwdValidate) {
                    PasswordValidator.LEAST_8 -> inputPwd.setError(true, getString(R.string.errAtLeast))

                    PasswordValidator.NOT_MATCH_PATTERN -> inputPwd.setError(true, getString(R.string.errPasswordPatternMatch))

                    PasswordValidator.HAS_WHITE_SPACE -> inputPwd.setError(true, getString(R.string.errWhiteSpace))

                    PasswordValidator.SERIAL_CHAR -> inputPwd.setError(true, getString(R.string.errSerialChar))

                    else -> inputPwd.setError(false, null)
                }
            }
        } else {
            inputPwd.setError(true, getString(R.string.errPwdEmpty))
            btnNext.isEnabled = false
            return
        }

        if (inputCheck.getText().isNotEmpty()) {
            matched = checkPasswordMatch(pwd, checkPwd)
            if (!matched) {
                inputCheck.setError(true, getString(R.string.errPasswordNotMatched))
            } else {
                inputCheck.setError(false, null)
            }
        } else {
            inputCheck.setError(true, getString(R.string.errCheckEmpty))
            btnNext.isEnabled = false
            return
        }

        if (aliasValidate == 0 && matched && pwdValidate == 0)
            btnNext.isEnabled = true
        else
            btnNext.isEnabled = false
    }

    private fun createWallet(): Array<String> {
        if (vm.getCoinType().value == CreateWalletViewModel.CoinType.ICX)
            return KeyStoreUtils.generateICXKeystore(inputPwd.getText())
        else
            return KeyStoreUtils.generateEtherKeystore(inputPwd.getText())
    }

    interface OnStep2Listener {
        fun onStep2Done()

        fun onStep2Back()
    }

    companion object {

        private val TAG = CreateWalletStep2Fragment::class.java.simpleName

        fun newInstance(): CreateWalletStep2Fragment {
            return CreateWalletStep2Fragment()
        }
    }
}// Required empty public constructor
